Access reviews tutorial (#11247)
microsoftgraph/microsoft-graph-docs: concepts/tutorial-accessreviews-M365group.md

---

> title: "Using the Access Reviews API for your M365 groups"

Reviewer: ```suggestion
title: "Create and read access reviews for your M365 groups using Microsoft Graph APIs."
```

description: "Use the Access Reviews API to review access to your M365 groups"
author: "isabelleatmsft"
localization_priority: Normal
ms.prod: "governance"
---

> # Tutorial: Use the Access Reviews API to review access to all your M365 groups with external identities

Reviewer: ```suggestion
# Tutorial: Create and read access reviews for M365 groups with guest users using the Microsoft Graph API.
```

> In this tutorial, you will use Graph Explorer to create and read access reviews that targets all M365 Groups with external identities in the tenant. To achieve this, you'll first use Azure AD B2B to invite and create an external identity in your tenant. Then, you'll add this external identity to your M365 group prior to creating and reading the access review.

Reviewer: ```suggestion
In this tutorial, you use Microsoft Graph Explorer to create and read access reviews that targets all M365 Groups with guest users in the tenant. To achieve this, you first use Azure AD B2B to invite and create a guest user in your tenant. Then, you add this guest user to your M365 group prior to creating and reading the access review.
```

## Prerequisites

To complete this tutorial, you need the following resources and privileges:

+ A working Azure AD tenant with an Azure AD Premium P2 or EMS E5 license enabled. 

> + An account in a different Azure AD tenant or a social identity that you can invite as an external identity (B2B user).

Reviewer: ```suggestion
+ An account in a different Azure AD tenant or a social identity that you can invite as a guest user (B2B user).
```

Author: Suggestion accepted and corrected in the remote branch.

> + Log in to [Microsoft Graph Explorer](https://developer.microsoft.com/graph/graph-explorer)  as a user in a global administrator role. 

Reviewer: ```suggestion
+ Sign in to [Microsoft Graph Explorer](https://developer.microsoft.com/graph/graph-explorer)  as a user in a global administrator role. 
```

> + Permissions:

Reviewer: I don't think this list is needed because you are listing them in step 3.

Author: Noted.

  + To call the invitation APIs to invite a user, consent to *User.Invite.All* permission.
  + To call the Access Reviews APIs, consent to the *AccessReview.ReadWrite.All* permission.
  + To create the M365 group, and delete it later after completing the tutorial, consent to the *Group.ReadWrite.All* permission.
  + To remove the guest user from the tenant after completing the tutorial, consent to the *Directory.AccessAsUser.All* permission.

To consent to the required permissions in Microsoft Graph Explorer:
1. Click **Sign in to Graph Explorer** and sign in using the account that has a global administrator role.

> 2. Select the settings gear icon to the right of the user account details, and then select **Select permissions**.

Reviewer: ```suggestion
2. Select the settings icon to the right of the user account details, and then choose **Select permissions**.
```

   
   ![Select the Microsoft Graph permissions](../images/../concepts/images/tutorial-accessreviews-api/settings.png)

> 3. Scroll through the list of permissions to find the *AccessReview.ReadWrite.All*, *Directory.AccessAsUser.All*, *Group.ReadWrite.All*, and *User.Invite.All* permissions. To consent to the permission, select it first, then select **Consent**.

Reviewer: These permissions won't be visible unless the category is expanded. For example, to get to the AccessReview.ReadWrite.All permission, the user must first expand **AccessReview (3)**. Suggestion: 

Scroll through the list of permissions to these permissions:
- AccessReviews (3), expand and then select AccessReviews.ReadWrite.All
- Group (2), expand and then select Group.ReadWrite.All
- User (8), expand and then select User.Invite.All
User (8), expand and then select User.ReadWrite.All

Author: Confirming about Directory.AccessAsUser.All - I can't delete the guest user with the other permissions. I'll need another permission for that.

   
   ![Consent to the Microsoft Graph permissions](../images/../concepts/images/tutorial-accessreviews-api/consentpermissions.png)

>    >[!NOTE]

Reviewer: Please remove the note. Suggestion:

4. Select **Consent**, and then select **Accept** to accept the consent of the permissions. You do not need to consent on behalf of your organization for these permissions.

   While consenting to the permissions, in the pop window, do not consent on behalf of your organization because these permissions are not needed by other users. Select **Accept** to accept the consent of the permissions.

## Step 1: Invite a new person (an external identity) into your tenant as a B2B guest

### Request

> Run the following request to invite an external user with the email address **john.doe@outlook.com** to your tenant.

Reviewer: Please move this sentence above the Request header.

> Run the following request to invite an external user with the email address **john.doe@outlook.com** to your tenant.

Reviewer: ```suggestion
Invite a guest user with the email address of **john.doe@outlook.com** to your tenant.
```

}
```

### Response

```http
{
    "@odata.context": "https://graph.microsoft.com/beta/$metadata#invitations/$entity",
    "id": "3043523a-7a7d-4aec-b9c7-141d9c277559",
    "inviteRedeemUrl": "https://login.microsoftonline.com/redeem?rd=https%3a%2f%2finvitations.microsoft.com%2fredeem%2f%3ftenant%3df9a44fab-6764-4166-aac0-c8aca790a100%26user%3d3043523a-7a7d-4aec-b9c7-141d9c277559%26ticket%3dc7qakt671UqG9jN4csqlXOvT9JipN2C8o4eGbh%252bvLWM%253d%26ver%3d2.0",
    "invitedUserDisplayName": "John Doe (Tailspin Toys)",
    "invitedUserType": "Guest",
    "invitedUserEmailAddress": "john.doe@outlook.com",
    "sendInvitationMessage": false,
    "resetRedemption": false,
    "inviteRedirectUrl": "https://myapps.microsoft.com/",
    "status": "PendingAcceptance",
    "invitedUserMessageInfo": {
        "messageLanguage": null,
        "customizedMessageBody": null,
        "ccRecipients": [
            {
                "emailAddress": {
                    "name": null,
                    "address": null
                }
            }
        ]
    },
    "invitedUser": {
        "id": "baf1b0a0-1f9a-4a56-9884-6a30824f8d20"
    }
}
```

> Take note of the invited user's id from the response. In this case, john.doe@outlook.com has been assigned user `id` `baf1b0a0-1f9a-4a56-9884-6a30824f8d20` in your tenant.

Reviewer: ```suggestion
Record the **id** of the guest user from the response to be used later in this tutorial.
```

> ## Step 2: Create a new M365 group and add the external guest

Reviewer: ```suggestion
## Step 2: Create a new M365 group and add the guest user
```

In this step:
1. Create a new M365 group named **Feelgood marketing campaign** and add `john.doe@outlook.com` to the group.
2. Assign yourself as the group owner.

> 3. Add john.doe@outlook.com as a group member. Their access to the group will be the subject of review by you, the group owner.

Reviewer: ```suggestion
3. Add john.doe@outlook.com as a group member. Their access to the group is the subject of review by you, the group owner.
```

### Request
In this call, replace:

> + `cdb555e3-b33e-4fd5-a427-17fadacbdfa7` with your `id`.

Reviewer: Where has the user been instructed in how to obtain their own id?

> + `baf1b0a0-1f9a-4a56-9884-6a30824f8d20` with the `id` john.doe@outlook.com has been assigned in your tenant.

Reviewer: Use the **id** that you recorded for the invited user.

```http
POST https://graph.microsoft.com/beta/groups
Content-Type: application/json

{
    "description": "Feelgood Marketing Campaign with external partners and vendors.",
    "displayName": "Feelgood Marketing Campaign",
    "groupTypes": [
        "Unified"
    ],
    "mailEnabled": true,
    "mailNickname": "FeelGoodCampaign",
    "securityEnabled": true,
    "owners@odata.bind": [
        "https://graph.microsoft.com/beta/users/cdb555e3-b33e-4fd5-a427-17fadacbdfa7"
    ],
    "members@odata.bind": [
        "https://graph.microsoft.com/beta/users/baf1b0a0-1f9a-4a56-9884-6a30824f8d20"
    ]
}
```

### Response

```http
{

>     "@odata.context": "https://graph.microsoft.com/beta/$metadata#groups/$entity",

Reviewer: In other tutorials, we have a note in the introduction that says that example responses may be shortened. You might want to consider shortening this response a bit and only including properties that are needed for the task.

Note: The response objects shown in this tutorial might be shortened for readability.

    "displayName": "Feelgood Marketing Campaign",
    "expirationDateTime": null,
    "groupTypes": [
        "Unified"
    ],
    "infoCatalogs": [],
    "isAssignableToRole": null,
    "isManagementRestricted": null,
    "mail": "FeelGoodCampaign@contoso.com",
    "mailEnabled": true,
    "mailNickname": "FeelGoodCampaign",
    "membershipRule": null,
    "membershipRuleProcessingState": null,
    "onPremisesDomainName": null,
    "onPremisesLastSyncDateTime": null,
    "onPremisesNetBiosName": null,
    "onPremisesSamAccountName": null,
    "onPremisesSecurityIdentifier": null,
    "onPremisesSyncEnabled": null,
    "preferredDataLocation": null,
    "preferredLanguage": null,
    "proxyAddresses": [
        "SMTP:FeelGoodCampaign@contoso.com"
    ],
    "renewedDateTime": "2021-02-10T16:55:34Z",
    "resourceBehaviorOptions": [],
    "resourceProvisioningOptions": [],
    "securityEnabled": true,
    "securityIdentifier": "S-1-12-1-1504404522-1311909750-4154419355-1786739620",
    "theme": null,
    "visibility": "Public",
    "writebackConfiguration": {
        "isEnabled": null,
        "onPremisesGroupType": null
    },
    "onPremisesProvisioningErrors": []
}
```

> ## Step 3: Create an access review for all M365 groups with external identities

Reviewer: ```suggestion
## Step 3: Create an access review for all M365 groups with guest users
```

> You have an M365 group and it has an external identity. Now, create a recurring access review series for all M365 groups with external guests. This means that for all M365 groups with external guests, you will schedule a periodic review of the guests' access to the M365 group.

Reviewer: ```suggestion
You have an M365 group and it has a guest user. You can then create a recurring access review series for all M365 groups with guest users. This means that for all M365 groups with guest users, you schedule a periodic review of the guests' access to the M365 group.
```

Author: Changes made to remote branch.

> The access review series will have the following settings:

Reviewer: ```suggestion
The access review series uses the following settings:
```

> + Its a recurring access review, reviewed quarterly.

Reviewer: ```suggestion
+ It's a recurring access review and reviewed quarterly.
```

> + The group owners review the external identity's continued access.

Reviewer: ```suggestion
+ The group owners review the continued access of the guest user.
```

> + The review scope is limited to M365 groups with **Guest users** (external identities) only.

Reviewer: ```suggestion
+ The review scope is limited to M365 groups with **Guest users** only.
```

> + A backup reviewer. This can be a fallback user or a group that can review the access in case the group doesn't have any owners assigned even though its group owners are the access review reviewers. 

Reviewer 2: cut "even though its group owners are the access review reviewers"

> + **autoApplyDecisionsEnabled** is set to `true`. In this case, once the reviewer completes the access review, their decisions are applied automatically. If not enabled, a user must, after the review completes, apply the decisions manually.

Reviewer 2: close but not exactly - the decisions are applied once the access review duration ends, regardless of whether the reviewer completed the review or not

> + Apply **removeAccessApplyAction** action to denied guest users. This removes the denied guest user's membership to the group, though they'll still be able to sign in to your tenant.[Learn more about the implications of this action](/azure/active-directory/governance/create-access-review#create-one-or-more-access-reviews).

Reviewer: ```suggestion
+ Apply **removeAccessApplyAction** action to denied guest users. This removes the membership in the group of the denied guest. The guest user can still sign in to your tenant.
```

> + Apply **removeAccessApplyAction** action to denied guest users. This removes the denied guest user's membership to the group, though they'll still be able to sign in to your tenant.[Learn more about the implications of this action](/azure/active-directory/governance/create-access-review#create-one-or-more-access-reviews).

Reviewer: Links that send the user out of the tutorial should be included in the See Also section, not the body of the tutorial.

> For more information about the parameters in the request body of these examples, see the [Access Reviews API Reference](/graph/api/resources/accessreviewsv2-root?view=graph-rest-beta&preserve-view=true).

Reviewer: This should be moved to the See Also section at the end of the tutorial.

### Request
In this call, replace the following:

> + Value of **displayName** with your preferred value.

Reviewer: Since this is only a learning experience and not for production, we should just provide a display name to use.

> + `c9a5aff7-9298-4d71-adab-0a222e0a05e4` with the `id` of the user you are designating as a backup reviewer.

Reviewer: Are we assuming that there are user accounts already created that the customer will know about using? I think it's better to have a section up front that creates the test user accounts needed for the tutorial and deletes them when done.

Author: Adding this step.

            "queryType": "MicrosoftGraph",
            "queryRoot": null
        }
    ],
    "settings": {
        "mailNotificationsEnabled": true,
        "reminderNotificationsEnabled": true,
        "justificationRequiredOnApproval": true,
        "defaultDecisionEnabled": true,
        "defaultDecision": "Approve",
        "instanceDurationInDays": 0,
        "autoApplyDecisionsEnabled": true,
        "recommendationsEnabled": true,
        "recurrence": {
            "pattern": {
                "type": "absoluteMonthly",
                "interval": 3,
                "month": 0,
                "dayOfMonth": 0,
                "daysOfWeek": [],
                "firstDayOfWeek": "sunday",
                "index": "first"
            },
            "range": {
                "type": "numbered",
                "numberOfOccurrences": 0,
                "recurrenceTimeZone": null,
                "startDate": "2021-02-10",
                "endDate": "2022-12-21"
            }
        },
        "applyActions": [
            {
                "@odata.type": "#microsoft.graph.removeAccessApplyAction"
            }
        ]
    }
}
```

> This is an access review definition. You can list and read this and any other existing definitions by running `GET` on `https://graph.microsoft.com/beta/identityGovernance/accessReviews/definitions`.

Reviewer: Links should be in the See Also section.

## Step 4: List instances of the access review

> The following query lists all instances of the access review definition. If your test tenant contains other M365 groups with external guests, this request will return more than 1 instance.

Reviewer 2: "this request will return an instance for every M365 group with guest users in the tenant"

### Request

> In this call, replace `c22ae540-b89a-4d24-bac0-4ef35e6591ea` with the `id` of a definition returned in Step 3.

Reviewer: Should be more specific about which id this is.

```http
GET https://graph.microsoft.com/beta/identityGovernance/accessReviews/definitions/c22ae540-b89a-4d24-bac0-4ef35e6591ea/instances
```

### Response

> In this response, te scope includes a group of `id` `59ab642a-2776-4e32-9b68-9ff7a47b7f6a`. This is the **Feelgood marketing campaign** group created in Step 2.

Reviewer: ```suggestion
In this response, the scope includes a group of `id` `59ab642a-2776-4e32-9b68-9ff7a47b7f6a`. This is the **Feelgood marketing campaign** group created in Step 2.
```

```http
HTTP/1.1 200 OK
Content-type: application/json

{
    "@odata.context": "https://graph.microsoft.com/beta/$metadata#identityGovernance/accessReviews/definitions('c22ae540-b89a-4d24-bac0-4ef35e6591ea')/instances",
    "value": [
        {
            "id": "6392b1a7-9c25-4844-83e5-34e23c88e16a",
            "startDateTime": "2021-02-10T17:00:36.96Z",
            "endDateTime": "2021-02-10T17:00:36.96Z",
            "status": "Applied",
            "scope": {
                "query": "/groups/59ab642a-2776-4e32-9b68-9ff7a47b7f6a/members/microsoft.graph.user/?$count=true&$filter=(userType eq 'Guest')",
                "queryType": "MicrosoftGraph"
            }
        }
    ]
}
```

> In this response, the **status** of the access review instance is marked as `Applied` because the current review cycle has ended. Because this is a quarterly review, every 3 months time, a new review instance is created and you—the reviewer—can apply new decisions.

Reviewer 2: for the purpose of the tutorial can we change the status to InProgress?

## Step 5: Get decisions

> You are interested in the decisions taken for the instance of an access review..

Reviewer: ```suggestion
Get the decisions taken for the instance of an access review.
```

In this call, replace `59ab642a-2776-4e32-9b68-9ff7a47b7f6a` with the `id` of your **Feelgood marketing campaign** M365 group.

```http
DELETE https://graph.microsoft.com/beta/groups/59ab642a-2776-4e32-9b68-9ff7a47b7f6a
```

#### Response

```http
HTTP/1.1 204 No Content
Content-type: text/plain
```

### Delete the access review definition

In this call, replace `c22ae540-b89a-4d24-bac0-4ef35e6591ea` with the `id` of your access review definition. Since the access review schedule definition is the blueprint for the access review, deleting the definition will remove the settings, instances, and decisions associated with the access review.

#### Request
```http
DELETE https://graph.microsoft.com/beta/identityGovernance/accessReviews/definitions/c22ae540-b89a-4d24-bac0-4ef35e6591ea
```

#### Response
```http
HTTP/1.1 204 No Content
Content-type: text/plain
```
### Remove the guest user

In this call, replace `baf1b0a0-1f9a-4a56-9884-6a30824f8d20` with the `id` of the guest user, john.doe@outlook.com.

#### Request
```http
DELETE https://graph.microsoft.com/beta/users/baf1b0a0-1f9a-4a56-9884-6a30824f8d20
```

#### Response
```http
HTTP/1.1 204 No Content
Content-type: text/plain
```

> Congratulations! You have created an access review for the M365 group and the group owners have approved access for the current review cycle. You have also triggered a scheduled quarterly re-evaluation and re-attestation of access for all guest users in M365 groups. The group owners will review access during these cycles, choosing either to approve or deny access.

Reviewer 2: "Congratulations! You have created an access review scheduled quarterly for the evaluation and attestation of access for all guest users in M365 groups. The group owners will review access during these cycles, choosing either to approve or deny access."

## See also

> 

Reviewer: A link to all API resources used in the tutorial should be included in this section.

---

> title: "Tutorial: Use the access reviews API to review access to all your Microsoft 365 groups with guest users - Microsoft Graph"

Reviewer 3: Same comments here regarding the title and description. Can we make the title more concise? Is the "with guest users" part something that can be removed?

Author: Unfortunately, the scope is very particular to guest users. How about "Use the access reviews API to review guest access to your Microsoft 365 groups"

description: "Use the access reviews API to review access to your Microsoft 365 groups"
author: "FaithOmbongi"
localization_priority: Normal
ms.prod: "governance"
---

# Tutorial: Use the access reviews API to review access to all your Microsoft 365 groups with guest users

> In this tutorial, you will use Graph Explorer to create and read access reviews that targets all Microsoft 365 Groups with guest users in the tenant. To achieve this, you'll first use Azure AD B2B to invite and create an guest user, also referred to as an external identity, in your tenant. Then, you'll add this guest user to your Microsoft 365 group prior to creating and reading the access review.

Reviewer 3: Groups > groups

>[!NOTE]

> >The response objects shown in this tutorial might be shortened for readability. All the properties will be returned from an actual call.

Reviewer 3: For technical accuracy, we now remove the "All the properties..." sentence from this boilerplate text.

## Prerequisites

To complete this tutorial, you need the following resources and privileges:

+ A working Azure AD tenant with an Azure AD Premium P2 or EMS E5 license enabled. 
+ An account in a different Azure AD tenant or a social identity that you can invite as a guest user (B2B user).

> + Sign in to [Graph Explorer](https://developer.microsoft.com/graph/graph-explorer)  as a user in a global administrator role. 

Reviewer 3: This third step, is that something they need or just something they do as they complete the tutorial?

Author: It's a requirement that they run this as a user in a global admin role. Let me remove step 1 below since they'll have done that.

> + Permission - For this tutorial, you need the following delegated permissions: `User.Invite.All`, `AccessReview.ReadWrite.All`, `Group.ReadWrite.All`, `User.ReadWrite.All`.

Reviewer 3: Because we're using noun phrases for this list of prerequisites, this can just say: "The following delegated permissions:".

   
   ![Select the Microsoft Graph permissions](../images/../concepts/images/tutorial-accessreviews-api/settings.png)
   <!--:::image type="content" source="../images/../concepts/images/tutorial-accessreviews-api/settings.png" alt-text="Select the Microsoft Graph permissions":::-->

3. Scroll through the list of permissions to these permissions:
   + AccessReviews (3), expand and then select **AccessReviews.ReadWrite.All**.
   + Group (2), expand and then select **Group.ReadWrite.All**.
   + User (8), expand and then select **User.Invite.All** and **User.ReadWrite.All**.
   
   Select **Consent**, and then select **Accept** to accept the consent of the permissions. You do not need to consent on behalf of your organization for these permissions.
   
   ![Consent to the Microsoft Graph permissions](../images/../concepts/images/tutorial-accessreviews-api/consentpermissions_M365.png)
   <!--:::image type="content" source="../images/../concepts/images/tutorial-accessreviews-api/consentpermissions_M365.png" alt-text="Consent to the Microsoft Graph permissions":::-->

## Step 1: Create a test user in your tenant

> Create a new user and record their **id**. You will assign them a role in the access review later in this tutorial. 

Reviewer 3: Do you mean to refer to an id property, or just a user ID?

### Request

```http
POST /users
Content-Type: application/json

{
    "accountEnabled": true,
    "displayName": "Aline Dupuy",
    "mailNickname": "AlineD",
    "userPrincipalName": "AlineD@contoso.com",
    "passwordProfile": {
        "forceChangePasswordNextSignIn": true,
        "password": "xWwvJ]6NMw+bWH-d"
    }
}
```

### Response

```http
HTTP/1.1 201 Created
Content-type: application/json

{
    "@odata.context": "https://graph.microsoft.com/beta/$metadata#users/$entity",
    "id": "c9a5aff7-9298-4d71-adab-0a222e0a05e4",
    "displayName": "Aline Dupuy",
    "userPrincipalName": "AlineD@contoso.com",
    "userType": "Member"
}
```

## Step 2: Invite a guest user into your tenant

> Invite a guest user with the email address **john.doe@outlook.com** to your tenant.

Reviewer 3: Technically you should use an approved fictitious name from the fictitious names list.

```

### Response

The following response shows the decision taken for the instance of the review.

```http
HTTP/1.1 200 OK
Content-type: application/json

{
    "@odata.context": "https://graph.microsoft.com/beta/$metadata#identityGovernance/accessReviews/definitions('c22ae540-b89a-4d24-bac0-4ef35e6591ea')/instances('6392b1a7-9c25-4844-83e5-34e23c88e16a')/decisions",
    "@odata.count": 1,
    "value": [
        {
            "id": "0e76ee07-b4c6-469e-bc9d-e73fc9a8d660",
            "accessReviewId": "6392b1a7-9c25-4844-83e5-34e23c88e16a",
            "reviewedDateTime": "2021-02-10T17:06:26.147Z",
            "decision": "Approve",
            "justification": "",
            "appliedDateTime": null,
            "applyResult": "New",
            "recommendation": "Deny",
            "reviewedBy": {
                "id": "00000000-0000-0000-0000-000000000000",
                "displayName": "AAD Access Reviews",
                "userPrincipalName": "AAD Access Reviews"
            },
            "appliedBy": {
                "id": "00000000-0000-0000-0000-000000000000",
                "displayName": "",
                "userPrincipalName": ""
            },

>             "target": {

Reviewer 2: Small thing - the structure of the accessReviewDecisionItem changed just slightly to include a new property called principal. Principal will replace target in the future (we will deprecate target). Can you rerun your tutorial and get the response here updated to include the returned parameter principal?
